src/controllers/auth.js

```diff
@@ -1,5 +1,217 @@
+  }
+  // if student not exist return username
+  return username;
+}
+
 module.exports = {
-  signin: async () => {},
-  signInWithGmail: async () => {},
-  signup: async () => {},
+  signin: async (req, res) => {
+    const { email, password } = req.body;
+
+    // check username and password
+    if (!email && !password) {
+      return res
+        .status(400)
+        .json({ error: 'email and password fields are required' });
+    }
+
+    try {
+      // find student from email
+      const student = await StudentModel.findOne({ email });
+
+      if (!student) {
+        return res.status(400).json({ error: 'wrong email or password' });
+      }
+      if (student.isSignUpWithGoogle) {
+        // student should login via using gmail.
+        // because there is not any password in database for this student.
+        // it requires google oauth verification.
+        return res.status(400).json({
+          error: 'This account is linked to Google. Please log in via Google',
+        });
+      }
+
+      // compare sent password to saved one
+      const result = await bcrypt.compare(password, student.passwordHash);
+      if (!result) {
+        return res.status(400).json({ error: 'wrong email or password' });
+      }
+
+      // generate jwt token and send it to client with student information in payload
+      const token = generateAccessToken({ student });
+      return res.status(200).json(token);
+    } catch (err) {
+      return res.status(400).json(err);
+    }
+  },
+  signup: async (req, res) => {
+    const { firstName, lastName, password, avatar, email, isTutor, subjects } =
+      req.body;
+
+    // password are required,
+    // length should longer than 6 characters
+    if (!password || password.length < 6) {
+      return res
+        .status(400)
+        .json({ error: 'password should longer than 6 characters' });
+    }
+
+    // Check existing student with given email
+    try {
+      const existingStudent = await StudentModel.findOne({ email });
+      if (existingStudent) {
+        return res
+          .status(400)
+          .json({ error: `${email}: already used for signup` });
+      }
+    } catch (error) {
+      return res.status(400).json({ error });
+    }
+
+    // hash password before persistion
+    const passwordHash = await bcrypt.hash(password, 10);
+
+    // get a unique username
+    const username = await getUniqueUserName(
+      firstName.trim() + lastName.trim()
+    );
+
+    try {
+      // Creating student field
+      const student = new StudentModel({
+        firstName,
+        lastName,
+        username,
+        avatar,
+        passwordHash,
+        email,
+        isTutor,
+      });
+
+      if (subjects) {
+        student.subjects = subjects;
+      }
+
+      const newStudent = await student.save();
+
+      // Generate jwt token
+      const token = generateAccessToken({ student: newStudent });
+
+      return res.status(201).json(token);
```

Comment:
amazing work @emirsagit. I just want to mention that I added an attribute "user" to the request parameter "req" using "cookie parser",  and assign to it the login token.  We can also use it for authorization in every request. I wonder if you implemented it a better way. We can discuss this issue in the next meeting.

Reply:
Ehab, thank you for reviewing. I do not totally sure, but I think the flow is like that in this implementation. Backend sent the JWT token to the client after successful login. Then the client gets the user information from the token and saves it to somewhere like a cookie, local storage, etc. Then the client will send the token to the backend on each request. We need to decide it or we can get the token on the backend searching on the middleware-like body.token || header.token etc. We can discuss it in the next meeting.